### DISCUSSION TITLE
Welcome to UnityCatalog Discussions!

### DISCUSSION THREAD

**Discussion by dennyglee** [unitycatalog/unitycatalog#208] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.  Any serious question is fair game (though let's try to stay on topic for UC otherwise I'll talk about coffee too much)
  * Share ideas. Share thoughts. Yes, we will be sharing the roadmap soon - meanwhile, we are updating the [milestones](https://github.com/unitycatalog/unitycatalog/milestones)
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by MrPowers** [comment]
Hi everyone, my name is Matthew Powers and I'm so excited to collaborate with the Unity Catalog community on this amazing project!

I've been writing Spark, pandas, Polars, and Delta Lake code for years and have accumulated tons of localhost datasets.  I immediately forget where all my datasets are stored and can never find them again!  One of my immediate goals is to create a script that generates a bunch of localhost datasets that are neatly organized in a Unity Catalog.  I hope to access `h2o.groupby.billion_rows` rather than searching my computer every time.

I am also excited about collaborating with the community to build beautiful docs!

**Comment by ognis1205** [comment]
Hi folks, I am Shingo, a developer based in Tokyo.

As of [13/07/2024], my primary focus in this community is developing the UI/UX layer for the Unity Catalog OSS, such as web UI applications. If we aim to democratize Data + AI for people from various backgrounds in this community, addressing the UI/UX is also essential. Please feel free to reach out to me whenever you want. Since I live in Tokyo, it might be easier to connect with me during East Asia time.

**Comment by avriiil** [comment]
Hi folks, my name is Avril Aysha and I'm a technical writer and devrel. I'm really excited to be joining this growing open source community. I will be contributing through working on the docs and other forms of content to spread the word and get more folks excited about Unity Catalog's awesome features! Please feel free to reach out anytime :)

**Comment by Buktoria** [comment]
Hi everyone! I'm Vicky and I'm a engineer, turned engineering manager, turned product manager. 

My background is with all things Iceberg and Kafka. I'm getting up to speed with Delta. The pain of having to manage different kinds of data assets was very real for me in my pervious life. I'm so excited that there is finally a project to tackle the governance and usability story of engines across formats. I'm happy to be a part of the Unity Catalog OSS community!

**Comment by michelleon** [comment]
Hi everyone! I'm Michelle, a product manager based in San Francisco! Excited to work with you all -- on the docs, roadmap, and growing the community and integrations.

In particular, most energized to work with our amazing partners to grow our integrations ecosystem and work on bringing some exciting new things to Unity Catalog.

**Comment by cakerly-lf** [comment]
Hi all! My name is Carly, a communications manager based in St. Petersburg, FL. Very excited to work with the community to boost Unity Catalog content + events.

**Comment by narausch** [comment]
From Juan-Luis Cano: 
 
We have been exploring Unity Catalog and unitycatalog-ui for a few days and we are still unsure in what way we should use it.
 
Part of it might be because there aren't many integrations yet. The DuckDB one is not yet a high priority https://github.com/duckdb/uc_catalog/issues/8, we are waiting on the Spark one to arrive, and I haven't seen any mentions on other Python dataframe libraries like Polars or Pandas.
 
Say that I already have my data as Delta Tables. Is UC an ancillary system that helps visualise the existing datasets + volumes + functions, and nothing in the code should change? Alternatively, are we expecting tighter integration with dataframe libraries in the future in a way that we will retrieve the data *through* UC, hence mostly programmatic access? Are both of these correct? Something else that I'm missing?
 
To clarify, docs and blog posts on the REST API are probably useful for tool-makers, but I contend that they don't tell the full story on how *end users* are supposed to use this, now or in the future.

**Reply by dennyglee** [reply]
Hi Juan-Luis Cano (and thanks @narausch for pasting this in this discussion), you’ve asked some great questions and let me try to address each of your questions. 

> We have been exploring Unity Catalog and unitycatalog-ui for a few days and we are still unsure in what way we should use it 

The key tenet of Unity Catalog is that there is centralized governance for managing and organizing data and AI assets across your organization.  As you build more and more data and AI assets (e.g., tables and models) it can get unwieldy in terms of data discoverability (e.g., what tables do I have, what models do I have) as well as governance (e.g., can person X get access to model A version 2 to be used against table Y version 20).  These are just examples of the problem but the idea is that Unity Catalog simplifies the data and AI asset management so it’s easier for cross team collaboration.  Just as important is that this cross-team collaboration is done with governance in mind - team C can access these set of tables and models, team D can access another set of tables and models.  Even in these “simplified” examples, this can get quite difficult to manage - and this is what Unity Catalog aims to solve.  To your point about UnityCatalog-UI, we wanted to provide an intuitive user experience to understand what data and AI assets you have and how to ensure access rights and asset discoverability (e.g., what tables and models do I have) for your users and groups.

> Part of it might be because there aren't many integrations yet. The DuckDB one is not yet a high priority https://github.com/duckdb/uc_catalog/issues/8, we are waiting on the Spark one to arrive, and I haven't seen any mentions on other Python dataframe libraries like Polars or Pandas.

Very true but only because as of this writing we’re 0.1.  We have a number of partners that are working on integrations and during the Unity Catalog bi-weekly Community Meetings (watch them [here](https://youtube.com/@unitycatalog-io)) we’ve already had a number of integration showcases.  As noted in the 0.2, we’re all working together on various integrations including DuckDB and Spark like you mentioned.  Yes, I definitely see us working within the context of the Python ecosystem for DataFusion, Polars, Pandas and more.  We have numerous threads chatting on this and would love PRs if you’re interested :) 

> Say that I already have my data as Delta Tables. Is UC an ancillary system that helps visualise the existing datasets + volumes + functions, and nothing in the code should change?

UC will certainly help you visualize the existing datasets, volumes, functions, and models (including the aforementioned Delta tables).  In terms of how your code could change, UC is designed to be transparent but it is important to note that as you add more governance around your assets, you _may_ want to modify code.  It is important to note here that this isn’t specific to UC, I’m just calling out the nature of governance that there are times you will want to change the code (for querying, processing, etc.) to reflect the governance you want to apply.

> Alternatively, are we expecting tighter integration with dataframe libraries in the future in a way that we will retrieve the data through UC, hence mostly programmatic access?

Yes, absolutely!  This is the yin/yan of Unity Catalog - giving you the ability to visualize your assets and at the same time being to simplify programmatic access.  There are times where visualization is preferred and there are times that programmatic access makes most sense.  From my personal experience of data management and governance going way back to SQL Server 2000 (yeah, its been awhile for me), you’re going to need both and will choose the medium that makes most sense.  So you’re right - these are both correct! :) 

> To clarify, docs and blog posts on the REST API are probably useful for tool-makers, but I contend that they don't tell the full story on how end users are supposed to use this, now or in the future.

Completely agree with your assertion that REST APIs don’t tell the full story.  A lot of our focus in these early days of Unity Catalog (as of this writing, we released 0.1 and we are progressing on 0.2) are around the APIs because this is the part we have to get right first.  We will build features/tools/visualzations based on these APIs so we need to make sure we establish a solid foundation.  This doesn’t mean things won’t change (after all, we all have fun with API versioning) but it’s a lot easier for us to build all of this on that solid foundation.

I hope this helps and glad to discuss more!  For that matter, glad to also dive into this during our next UC community meeting as well.  Thanks for taking the time to read this!

**Reply by astrojuanlu** [reply]
Original Juan-Luis here 😄 thanks a lot @dennyglee for the reply! (And @narausch for relaying my question here)

**Reply by dennyglee** [reply]
Awesome stuff @astrojuanlu !

**Comment by JuanPabloDiaz** [comment]
Hi everyone!

I'm Juan Díaz, a frontend developer based in Charlotte, NC. I'm excited to contribute to the Unity Catalog project, especially on improving the UI to make it more accessible and engaging. Looking forward to collaborating with this amazing community!

**Comment by ognis1205** [comment]
> Completely agree with your assertion that REST APIs don’t tell the full story. A lot of our focus in these early days of Unity Catalog (as of this writing, we released 0.1 and we are progressing on 0.2) are around the APIs because this is the part we have to get right first. We will build features/tools/visualzations based on these APIs so we need to make sure we establish a solid foundation. This doesn’t mean things won’t change (after all, we all have fun with API versioning) but it’s a lot easier for us to build all of this on that solid foundation.

Relating to this, I found there are several issues, PRs, and implementation plans progressing around the "client" implementation for the OpenAPI specification, with each being advanced as independent third-party libraries and implementation plans. Is there any plan to consolidate the design policies at some stage? For now, is it okay to proceed with the implementation according to each individual design philosophy?

 - https://github.com/unitycatalog/unitycatalog/pull/309
 - https://unitycatalog.slack.com/archives/C078UDLTVR7/p1722411967087619
 - https://github.com/unitycatalog/unitycatalog-ui/issues/57

**Reply by dennyglee** [reply]
Yes, as we progress, we should be working together on the agreed upon "core" YAML.  From there, there may be some understandable debate which set of code will be codegen vs. manually created to minimize these implementation issues.

**Reply by ognis1205** [reply]
Thank you for elaborating the situation :)

**Comment by erskine** [comment]
Howdy all! I'm Erskine Williams, a technical product manager with Hydrolix, out of Portland, Oregon. Our product can achieve up to a 50x improvement in compression of timeseries data over parquet with no sacrifice to query performance. We have written a spark connector (DataSourcev2 impl) for our binary file format, and we're looking to add Unity Catalog support to enable users to govern access to their data. We are eager to contribute back as needed to make it easy/easier to integrate arbitrary spark connectors into the Unity Catalog, but first, we may need some guidance & orientation.

**Reply by dennyglee** [reply]
Very nice to meet another PNWer @erskine  :) Looking forward to learning more about your connector, any chance you have a doc and/or GitHub location so we can review?

**Reply by erskine** [reply]
Thanks so much @dennyglee! You can see v1 of our integration [here](https://github.com/hydrolix/spark-connector). We have taken the car back into the shop to tune it up, and v2 should be coming out shortly, with enhanced support for predicate pushdown and other Spark goodies as well as Unity Catalog integration. It's also worth peeking at Clickhouse's spark connector [here](https://github.com/ClickHouse/spark-clickhouse-connector), since we are (loosely) built on Clickhouse. 

Most immediately, our fundamental question is how to delegate a sql query to our connector after we register our catalog with UC. I'm looking closely at [this implementation](https://books.japila.pl/unity-catalog-internals/spark-integration/#spark_catalog) for hints, but if the community here has a clear explanation for how we could register a non-delta spark-enabled custom connector with UC, that would help us immensely.

**Reply by dennyglee** [reply]
Oh got it - may have to wait just a few weeks as we're currently working on the spark connector per #231.  Once we get this in, this may provide a nice blueprint for your connector as well.

**Comment by dbverse** [comment]
Hi Team,

I've observed an interesting behavior when writing external tables to Unity Catalog (OSS) running locally, and I'm hoping to get some clarity.

When tables are created and registered via PySpark, they appear in the Unity Catalog UI, but without any column metadata (no column names or their datatypes are displayed). However, tables created using the uc bash commands show their metadata correctly in the Unity Catalog UI.

Is there a specific reason why spark-submit or general PySpark execution might not be creating this metadata in Unity Catalog, while the uc CLI does?

Below are snippets for both cases for reference:

### Case 1: Table Creation via PySpark (spark.sql)

Python

<pre><code>ddl_statement = f"""
    CREATE OR REPLACE TABLE demo.uc_emp.address
    (AddressID INT, AddressLine1 STRING, AddressLine2 STRING, City STRING, StateProvince STRING,
    CountryRegion STRING, PostalCode STRING, rowguid STRING, ModifiedDate TIMESTAMP)
    USING DELTA
    LOCATION '{delta_table_path}'
    COMMENT 'External table registered in Unity Catalog from existing Delta files';
    """
spark.sql(ddl_statement).show() 
</code></pre>

Observed Behavior: Table demo.uc_emp.address is registered, but its columns are not visible in the Unity Catalog UI.

### Case 2: Table Creation via uc Bash Command

Bash
<pre><code>
~/spark_env/unitycatalog$ bin/uc table create \
    --full_name demo.uc_emp.address_2 \
    --storage_location /home/processed/uc_address \
    --format delta \
    --columns "AddressID INT, AddressLine1 STRING, AddressLine2 STRING, City STRING, StateProvince STRING, CountryRegion STRING, PostalCode STRING, rowguid STRING, ModifiedDate TIMESTAMP"
</code></pre>

Observed Behavior: Table demo.uc_emp.address_2 is registered, and all column names and datatypes are correctly displayed in the Unity Catalog UI.

Any insights or guidance on this would be greatly appreciated! Thanks in advance.

**Reply by Kavan-M** [reply]
I have been running into a similar issue and believe it is connected so will reply here instead of creating a new discussion.

I have been running Unity Catalog OSS locally too and have been testing creation and registration of tables via pyspark. The tables get registered in the catalog, verified by the table listing using the CLI and querying using spark. Using table get on the CLI though I do not get the column metadata (empty), and also running table read it is not capable of reading the table. Using pyspark though everything works as it should, so I can retrieve information on the table and read it too (using spark.sql("SELECT * FROM schema.table").show())

What I notice is that in the STORAGE_LOCATION, the new tables have paths that seem to indicate a file, whereas if I do table get on one of the example files, the path ends in /. So for example, the example table numbers has a path of file:///unitycatalog/etc/data/external/unity/default/tables/numbers/, but tables I created get paths file:///path/to/table, despite trying to specify file:///path/to/table/. Do not know if this is of importance and haven't scoured the source code at all yet.

**Reply by dbverse** [reply]
Thanks for the insights. As I specified the path in 'bash' command (the command I had shared in the thread earlier), I don't see a difference in the paths when I tried 'describe extended' on both tables. I pointed to a similar path where I had created an external table using spark-submit. Yet, only the table created using the UC bash command had the column metadata.

**Comment by shekharsiva26** [comment]
Would there be a way to get a helm chart for unity catalog for a production ready deployment? Although it works locally a production ready scale helm chart will really help.

**Comment by haho16** [comment]
Hey, I would like to start a Q/A discussion about read/write capabilities of Spark connector for Databricks UC, but I think I need permission to do that. Is that correct?

**Comment by CristianPachacama** [comment]
Hi! I'm a Data Science & AI team lead working with data platforms and MLOps pipelines. 
Big fan of what you're building with UC — we're evaluating it as the core governance 
layer for our team's data and model assets.

I wanted to share two feature ideas that came up while thinking about where UC could go next:

---

### 💡 Idea 1 — Semantic Layer on top of Unity Catalog

**Problem:** UC does a great job with technical metadata, but there's no way to express 
what data *means* — domain concepts, typed relationships between assets, hierarchies. 
This is a real gap when building AI Agents that use the catalog as a knowledge base. 
Without semantic context, agents hallucinate more and reason less precisely.

**Proposal:** A Semantic Layer sitting *above* the current UC model — not replacing 
anything, just a higher-level abstraction where you can define domain concepts 
(`Customer`, `Order`), typed relationships (`Customer → placed → Order`), and 
constraints. Think of it as making UC **knowledge-graph-aware**. The immediate win 
is better grounding for AI Agents, but it also opens the door to semantic lineage 
and domain-driven governance.

---

### 💡 Idea 2 — Metadata Federation (explore before you ingest)

**Problem:** To understand what's in an external source system, you have to go there 
directly. There's no way to register and browse its structure from within UC without 
physically ingesting the data first — which leads teams to replicate entire datasets 
just to figure out what they actually need.

**Proposal:** A Metadata Federation mode in UC that lets you register external sources 
(operational DBs, SaaS platforms, legacy systems) as catalog objects — metadata only, 
no data movement. Just schemas, tables, columns, basic stats. Teams could then browse 
source structures natively in UC and decide *what* to ingest before building any pipeline. 
Different from Lakehouse Federation (which runs queries) — this is lighter, read-only, 
governance-first.

---

**Comment by jay-chayutapon** [comment]
Hi team, thank you for all your contributions. Could you please let me know when you plan to release the following features?

- RBAC
- Row level filters
- Column level masks
- Lineage
